Java implementation of Bull in a china shop (#2758)
cpinitiative/usaco-guide: solutions/usaco-640.mdx

id: usaco-640
source: USACO Bronze 2016 Open
title: Bull in a China Shop
author: Maggie Liu, David Zhang
---

> [Official Analysis](http://www.usaco.org/current/data/sol_bcs_bronze_open16.html)

Reviewer: ```suggestion
[Official Analysis (Java)](http://www.usaco.org/current/data/sol_bcs_bronze_open16.html)
```

					}
				}
			}
			s[i] = new int[]{left, right, top, bottom};
		}

		// try all the pieces and shifts to find the correct one
		for (int i = 1; i <= k; i++) {
			for (int j = i + 1; j <= k; j++) {
				for (int idx = s[i][3] - n + 1; idx <= s[i][2]; idx++) {
					for (int idy = s[i][1] - n + 1; idy <= s[i][0]; idy++) {
						for (int jdx = s[j][3] - n + 1; jdx <= s[j][2]; jdx++) {
							for (int jdy = s[j][1] - n + 1; jdy <= s[j][0]; jdy++) {
								boolean good = true;
								for (int x = 0; x < n; x++) {
									for (int y = 0; y < n; y++) {
										boolean ipiece = check(i, x + idx, y + idy);
										boolean jpiece = check(j, x + jdx, y + jdy);
										// two '#' are in the same place
										if (ipiece && jpiece) {	
											good = false;
											break;
										}
										// the result doesn't match the figurine
										if (pieces[0][x][y] != (ipiece || jpiece)) {
											good = false;
											break;
										}
									}
									if (!good) {
										break;
									}
								}
								if (good) {
									io.println(i + " " + j);
									System.exit(0);
								}
							}
						}
					}
				}
			}
		}
		io.close();

> 	}
> 	// check if a piece is in bounds and is '#'

Reviewer 2: ```suggestion
	}

	// check if a piece is in bounds and is '#'
```

	}
}

// check if a piece is in bounds and is '#'
bool check(int piece, int x, int y) {
	return x >= 0 && x < n && y >= 0 && y < n && pieces[piece][x][y];
}
```

</CPPSection>

<JavaSection>
```java
import java.util.*;
import java.io.*;

public class BCS {
	public static boolean pieces[][][];
	public static int n;
	public static void main(String[] args) throws IOException {
		Kattio io = new Kattio("bcs");
	
		pieces = new boolean[11][8][8];

> 		int[][] s = new int[11][4];

Reviewer 2: was thinking about this
how about you put:
11 in a variable called PIECE_NUM
dynamically initialize 8 with `n` instead (so `new boolean[PIECE_NUM][n][n]`)
and just do `new int[PIECE_NUM][]` instead of `new int[PIECE_NUM][4]`

		n = io.nextInt();
		int k = io.nextInt();

> 		int left, right, top, bottom;
> 		for (int i = 0; i <= k; i++) {
> 			left = n - 1;
> 			right = 0;
> 			top = n - 1;
> 			bottom = 0;

Reviewer 2: ```suggestion
		for (int i = 0; i <= k; i++) {
			int left = n - 1;
			int right = 0;
			int top = n - 1;
			int bottom = 0;
```

			for (int j = 0; j < n; j++) {
				String str = io.next();
				for (int l = 0; l < n; l++) {
					char c = str.charAt(l);

> 					pieces[i][j][l] = (c == '#');

Reviewer 2: parens not needed

					}
				}
			}
			s[i] = new int[]{left, right, top, bottom};
		}

		// try all the pieces and shifts to find the correct one
		for (int i = 1; i <= k; i++) {
			for (int j = i + 1; j <= k; j++) {
				for (int idx = s[i][3] - n + 1; idx <= s[i][2]; idx++) {
					for (int idy = s[i][1] - n + 1; idy <= s[i][0]; idy++) {
						for (int jdx = s[j][3] - n + 1; jdx <= s[j][2]; jdx++) {
							for (int jdy = s[j][1] - n + 1; jdy <= s[j][0]; jdy++) {
								boolean good = true;
								for (int x = 0; x < n; x++) {
									for (int y = 0; y < n; y++) {
										boolean ipiece = check(i, x + idx, y + idy);
										boolean jpiece = check(j, x + jdx, y + jdy);
										// two '#' are in the same place
										if (ipiece && jpiece) {	
											good = false;
											break;
										}
										// the result doesn't match the figurine
										if (pieces[0][x][y] != (ipiece || jpiece)) {
											good = false;
											break;
										}
									}
									if (!good) {
										break;
									}
								}
								if (good) {
									io.println(i + " " + j);
									System.exit(0);
								}
							}
						}
					}
				}
			}
		}
		io.close();
	}

> 	// check if a piece is in bounds and is '#'

Reviewer 2: ```suggestion
	/** checks if a piece location is in bounds and is '#' */
```

	}
}

// check if a piece is in bounds and is '#'
bool check(int piece, int x, int y) {
	return x >= 0 && x < n && y >= 0 && y < n && pieces[piece][x][y];
}
```

</CPPSection>

<JavaSection>
```java
import java.util.*;
import java.io.*;

public class BCS {
	public static boolean pieces[][][];
	public static int n;
	public static void main(String[] args) throws IOException {
		Kattio io = new Kattio("bcs");
	
		pieces = new boolean[11][8][8];
		int[][] s = new int[11][4];
		n = io.nextInt();
		int k = io.nextInt();

> 		char c;

Reviewer 2: don't declare `c` here when you're just using it in that input reading below

		int left, right, top, bottom;
		for (int i = 0; i <= k; i++) {
			left = n - 1;
			right = 0;
			top = n - 1;
			bottom = 0;
			for (int j = 0; j < n; j++) {
				String str = io.next();
				for (int l = 0; l < n; l++) {
					c = str.charAt(l);
					pieces[i][j][l] = (c == '#');
					// find the sides of the piece
					if (pieces[i][j][l]) {
						bottom = Math.max(bottom, j);
						top = Math.min(top, j);
						right = Math.max(right, l);
						left = Math.min(left, l);
					}
				}
			}
			s[i] = new int[]{left, right, top, bottom};
		}

		// try all the pieces and shifts to find the correct one
		for (int i = 1; i <= k; i++) {
			for (int j = i + 1; j <= k; j++) {
				for (int idx = s[i][3] - n + 1; idx <= s[i][2]; idx++) {
					for (int idy = s[i][1] - n + 1; idy <= s[i][0]; idy++) {
						for (int jdx = s[j][3] - n + 1; jdx <= s[j][2]; jdx++) {
							for (int jdy = s[j][1] - n + 1; jdy <= s[j][0]; jdy++)

> 							{

Reviewer 2: why's this brace on a separate line

								boolean good = true;
								for (int x = 0; x < n; x++) {
									for (int y = 0; y < n; y++) {
										boolean ipiece = check(i, x + idx, y + idy);
										boolean jpiece = check(j, x + jdx, y + jdy);
										// two '#' are in the same place
										if (ipiece && jpiece) {	
											good = false;
											break;
										}
										// the result doesn't match the figurine
										if (pieces[0][x][y] != (ipiece || jpiece)) {
											good = false;
											break;
										}
									}
									if (!good) {
										break;
									}
								}
								if (good) {
									io.println(i + " " + j);

> 									return;

Reviewer 2: _maybe_ `System.exit(0)` is more explicit?

</CPPSection>

<JavaSection>
```java
import java.util.*;
import java.io.*;

> class Main {

Reviewer 2: ```suggestion

public class BCS {
```

	public static boolean pieces[][][];
	public static int[][] s;
	public static int n, k;

> 	public static char c;

Reviewer 2: don't need to be global

	public static void main(String[] args) throws IOException {
		Kattio io = new Kattio("bcs");
		pieces = new boolean[11][8][8];
		s = new int[11][4];
		n = io.nextInt();
		k = io.nextInt();
		int left, right, top, bottom;
		for (int i = 0; i <= k; i++) {
			left = n - 1;
			right = 0;
			top = n - 1;
			bottom = 0;
			for (int j = 0; j < n; j++) {
				String str = io.next();
				for (int l = 0; l < n; l++) {
					c = str.charAt(l);
					pieces[i][j][l] = (c == '#');
					// find the sides of the piece
					if (pieces[i][j][l]) {
						bottom = Math.max(bottom, j);
						top = Math.min(top, j);
						right = Math.max(right, l);
						left = Math.min(left, l);
					}
				}
			}
			s[i] = new int[]{left, right, top, bottom};
		}

		// try all the pieces and shifts to find the correct one
		for (int i = 1; i <= k; i++) {
			for (int j = i + 1; j <= k; j++) {
				for (int idx = s[i][3] - n + 1; idx <= s[i][2]; idx++) {
					for (int idy = s[i][1] - n + 1; idy <= s[i][0]; idy++) {
						for (int jdx = s[j][3] - n + 1; jdx <= s[j][2]; jdx++) {

> 							for (int jdy = s[j][1] - n + 1; jdy <= s[j][0]; jdy++)

Reviewer 2: WHAT

Author: yeah, it was the solution from Maggie

							{
								boolean good = true;
								for (int x = 0; x < n; x++) {
									for (int y = 0; y < n; y++) {
										boolean ipiece = check(i, x + idx, y + idy);
										boolean jpiece = check(j, x + jdx, y + jdy);
										// two '#' are in the same place
										if (ipiece && jpiece) {	
											good = false;
											break;
										}
										// the result doesn't match the figurine
										if (pieces[0][x][y] != (ipiece || jpiece)) {
											good = false;
											break;
										}
									}
									if (!good) {
										break;
									}
								}
								if (good) {
									io.println(i + " " + j);
									return;
								}
							}
						}
					}
				}
			}
		}
		io.close();
	}
	// check if a piece is in bounds and is '#'
	static boolean check(int piece, int x, int y)

> 	{

Reviewer 2: opening brace on same line

```
</JavaSection>
<PySection>
```py
from sys import exit

with open('bcs.in') as read:

>     n, k = [int(i) for i in read.readline().split()]

Reviewer 2: crap man can you also change the indentation to tabs

Author: No worries, I'll fix it
